### ISSUE TITLE
Support remote URLs in Node.js and CLI

### ISSUE THREAD

**Issue by daniel-gato** [donmccurdy/glTF-Transform#452]
As per the specifications external URLs for assets are not forbidden in gltfs for textures and .bins.

Somehow, the cli utility is trying to fetch the url from the current path: 

```bash
error: ENOENT: no such file or directory, open '/Users/daniel/Desktop/https:/cdn.exmaple.io/dvpt-dg/shared_assets/files/000/001/465/9e3f8087b67dcaf4d1bed33bd23699b760637e6bf2e3a2c7883dc2e6933c9e04-optimised_jpg.jpg'
```

Am I missing something here?

I understand that `validate` would say :
```
Non-relative URI found: 'https://....jpg'.glTF Validator(NON_RELATIVE_URI)
```

But this shouldn't block the other cli commands.

Any thoughts?

**Comment by donmccurdy** [comment]
Currently this project has two I/O classes: NodeIO and WebIO. NodeIO doesn't make network requests (you generally don't want that in a server environment), and so the CLI doesn't make network requests either. I'm not completely sure the CLI should make network requests silently either, but certainly an opt-in flag (e.g. `--allow-remote`) would be a valid feature request. And in the meantime it should probably show a clearer error rather than trying to resolve the URL on disk. So I'll mark this as a feature request for now.

One option, in the meantime, is to do processing through the scripting API instead. For example:

```js
const path = require('path');
const fetch = require('node-fetch');

const json = await fetch('path/to/model.gltf').then((res) => res.json());
const resources = {};

for (const image in json.images) {
  const basename = path.basename(image.uri);
  resources[basename] = await fetch(image.uri);
  image.uri = basename;
}
for (const buffer in json.buffers) {
  ...
}

const document = new NodeIO().readJSON({json, resources});
```

**Comment by HexaField** [comment]
Hi there! I would love to support this issue. The XREngine team already uses this library and we are stuck not being able to load instanced GLTF models on our nodejs backend.

I believe there is a significant use case for loading GLTFs on a nodejs server, for things like user generated content, dynamic scene loading, server side 3d rendering & pixel streaming, various NFT functionality, the list goes on...

Please let us know how you think this could be best implemented and we will make swift work of it. I believe all that is needed is for the ability to pass a `URL` object into `fs.readFileSync` instead of a uri string, though I understand this may be more complicated when thinking about .gltf with internally referenced relative .bin paths.

**Comment by donmccurdy** [comment]
Thanks for reaching out, @HexaField! Let me think about this a bit more and get back to you. I'd like to see this supported, but `fs.readFileSync` doesn't have that capability. The solution would probably look something like a version of WebIO that uses [node-fetch](https://www.npmjs.com/package/node-fetch) instead of native Fetch API.

I'd like for the upcoming release (v2.0) to use async APIs for resource loading in both Web and Node.js — even when reading from disk that's likely better for performance in a server environment — so that's another upcoming change, which (fortunately) is also required for supporting HTTP requests in Node.js. 

And just to check – you're using the this package programmatically, not via CLI, is that correct?

Related:
- https://github.com/donmccurdy/glTF-Transform/issues/440

**Comment by hamzzam** [comment]
Yes, that is correct we are using the package programmatically. @donmccurdy

**Comment by donmccurdy** [comment]
Ok, I've merged #440 so there shouldn't be anything blocking the addition of async functionality like network requests in the I/O classes now. Would an API like this do what you're looking for?

```js
import fetch from 'node-fetch';
import { NodeIO } from '@gltf-transform/core';

const io = new NodeIO(fetch, fetchConfig);

io.setAllowHTTP(true); // default false

const document = await io.read('https://example.com/model.gltf');
const document = await io.read('local/path/to/model.gltf');
```

After support lands in NodeIO we could also provide a way to opt in for CLI users:

```shell
gltf-transform inspect input.gltf --allow-http
```

If that looks OK, it requires at least one main change. NodeIO currently uses `fs.readFile` to load files from disk, called in several of its methods. That'd need to be abstracted into a shared method that chooses whether to read from disk or to make an HTTP request depending on some context.  The current use of `path.resolve` might also need to be replaced — it's meant for resolving paths on disk! — with something more like WebIO currently does. If you're open to trying a PR for that I'd appreciate it! Let me know if you have any issues getting started.

**Comment by donmccurdy** [comment]
Aside — a word on security. Be careful about accepting user-generated content (UGC) from untrusted sources. Users can do evil things. Suppose you have a server endpoint that accepts user-generated GLBs, compresses them, and sends them back to the user. The user could put an "image" into the GLB with a "URL" pointing at some secret file from your server's disk, and NodeIO's `writeBinary` method would read that secret file and embed it into the resulting GLB. While the addition of a `setAllowHTTP(...)` method is convenient, this library doesn't know your use cases and application context, and where UGC is involved you'll probably want to be handling HTTP requests, checking URLs and paths on disk, and perhaps re-compressing images. Those risks aren't new with support for HTTP requests, but it does add a couple more attack vectors when enabled.

It might also be nice to expose an `io.setAllowFilesystem(bool)` method so that users can disable disk access if they don't need it.

**Comment by donmccurdy** [comment]
Fixed by @HexaField in https://github.com/donmccurdy/glTF-Transform/pull/462, with a few more changes to the API coming in #466. This will be published with v2.0.
